File: app/src/main/java/de/xikolo/controllers/course/SectionListAdapter.java
package de.xikolo.controllers.course;

import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.text.DateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import androidx.core.content.ContextCompat;
import androidx.core.view.ViewCompat;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import de.xikolo.R;
import de.xikolo.models.Section;
import de.xikolo.utils.DateUtil;
import de.xikolo.utils.DisplayUtil;
import de.xikolo.views.AutofitRecyclerView;
import de.xikolo.views.SpaceItemDecoration;

public class SectionListAdapter extends RecyclerView.Adapter<SectionListAdapter.SectionViewHolder> {

    public static final String TAG = SectionListAdapter.class.getSimpleName();

    private List<Section> sections;

    private FragmentActivity activity;

    private ItemListAdapter.OnItemClickListener itemClickListener;
    private OnSectionClickListener sectionClickListener;

    public SectionListAdapter(FragmentActivity activity, OnSectionClickListener moduleCallback,
                              ItemListAdapter.OnItemClickListener itemCallback) {
        this.activity = activity;
        this.sections = new ArrayList<>();
        this.sectionClickListener = moduleCallback;
        this.itemClickListener = itemCallback;
    }

    public void updateSections(List<Section> sections) {
        this.sections = sections;
        this.notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return sections.size();
    }

    @Override
    public SectionViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_section, parent, false);
        return new SectionViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final SectionViewHolder holder, int position) {
        final Section section = sections.get(position);

        holder.textTitle.setText(section.title);

        final ItemListAdapter itemAdapter = new ItemListAdapter(section, itemClickListener);
        holder.recyclerView.setAdapter(itemAdapter);
        holder.recyclerView.setHasFixedSize(false);
        holder.recyclerView.clearItemDecorations();
        holder.recyclerView.addItemDecoration(new SpaceItemDecoration(
                activity.getResources().getDimensionPixelSize(R.dimen.card_horizontal_margin) / 2,
                activity.getResources().getDimensionPixelSize(R.dimen.card_vertical_margin) / 2,
                false,
                new SpaceItemDecoration.RecyclerViewInfo() {
                    @Override
                    public boolean isHeader(int position) {
                        return false;
                    }

                    @Override
                    public int getSpanCount() {
                        return holder.recyclerView.getSpanCount();
                    }

                    @Override
                    public int getItemCount() {
                        return itemAdapter.getItemCount();
                    }
                }
        ));
        ViewCompat.setNestedScrollingEnabled(holder.recyclerView, false);

        if (section.hasAccessibleItems()) {
            contentAvailable(section, holder);
            itemAdapter.updateItems(section.getAccessibleItems());
        } else {
            contentLocked(section, holder);
        }
    }

    private void contentAvailable(final Section section, SectionViewHolder holder) {
        holder.progressBar.setVisibility(View.GONE);
        holder.viewModuleNotification.setVisibility(View.GONE);
        holder.viewHeader.setBackgroundColor(ContextCompat.getColor(activity, R.color.apptheme_section_header_bg));
        holder.textTitle.setTextColor(ContextCompat.getColor(activity, R.color.apptheme_section_header_text));

        TypedValue outValue = new TypedValue();
        activity.getTheme().resolveAttribute(android.R.attr.selectableItemBackground, outValue, true);
        holder.layout.setForeground(ContextCompat.getDrawable(activity, outValue.resourceId));
        holder.layout.setOnClickListener((view) -> sectionClickListener.onSectionClicked(section.id));

        if (section.hasDownloadableContent()) {
            holder.viewDownloadButton.setVisibility(View.VISIBLE);
            holder.viewDownloadButton.setOnClickListener((view) -> sectionClickListener.onSectionDownloadClicked(section.id));
        } else {
            holder.viewDownloadButton.setVisibility(View.GONE);
        }
    }

    private void contentLocked(Section section, SectionViewHolder holder) {
        holder.progressBar.setVisibility(View.GONE);
        holder.viewModuleNotification.setVisibility(View.VISIBLE);
        holder.viewHeader.setBackgroundColor(ContextCompat.getColor(activity, R.color.apptheme_section_header_bg_locked));
        holder.textTitle.setTextColor(ContextCompat.getColor(activity, R.color.apptheme_section_header_text_locked));

        holder.layout.setClickable(false);
        holder.layout.setForeground(null);
        holder.viewDownloadButton.setVisibility(View.GONE);
        if (section.startDate != null && DateUtil.isFuture(section.startDate)) {
            DateFormat dateOut;
            if (DisplayUtil.is7inchTablet(activity)) {
                dateOut = DateFormat.getDateTimeInstance(DateFormat.LONG, DateFormat.SHORT, Locale.getDefault());
            } else {
                dateOut = DateFormat.getDateTimeInstance(DateFormat.MEDIUM, DateFormat.SHORT, Locale.getDefault());
            }

            holder.textModuleNotification.setText(String.format(activity.getString(R.string.available_at),
                    dateOut.format(section.startDate)));
        } else {
            holder.textModuleNotification.setText(activity.getString(R.string.module_notification_no_content));
        }
    }

    public interface OnSectionClickListener {

        void onSectionClicked(String sectionId);

        void onSectionDownloadClicked(String sectionId);

    }

    static class SectionViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.container) FrameLayout layout;
        @BindView(R.id.textTitle) TextView textTitle;
        @BindView(R.id.recyclerView) AutofitRecyclerView recyclerView;
        @BindView(R.id.containerProgress) ProgressBar progressBar;
        @BindView(R.id.header) View viewHeader;

        @BindView(R.id.moduleNotificationContainer) View viewModuleNotification;
        @BindView(R.id.moduleNotificationLabel) TextView textModuleNotification;

        @BindView(R.id.downloadBtn) View viewDownloadButton;

        public SectionViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }

    }

}

File: app/src/main/java/de/xikolo/controllers/course/SectionListAdapter.kt
package de.xikolo.controllers.course

import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import de.xikolo.R
import de.xikolo.models.Section
import de.xikolo.utils.extensions.is7inchTablet
import de.xikolo.utils.extensions.isFuture
import de.xikolo.views.AutofitRecyclerView
import de.xikolo.views.DateTextView
import de.xikolo.views.SpaceItemDecoration
import java.text.DateFormat
import java.util.Locale

class SectionListAdapter(
    private val activity: FragmentActivity,
    private val sectionClickListener: OnSectionClickListener,
    private val itemClickListener: ItemListAdapter.OnItemClickListener
) : RecyclerView.Adapter<SectionListAdapter.SectionViewHolder>() {

    companion object {
        val TAG: String = SectionListAdapter::class.java.simpleName
    }

    private val sections: MutableList<Section> = mutableListOf()

    fun updateSections(sections: List<Section>) {
        this.sections.clear()
        this.sections.addAll(sections)
        this.notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return sections.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SectionViewHolder {
        val view = LayoutInflater
            .from(parent.context)
            .inflate(
                R.layout.item_section_list,
                parent,
                false
            )
        return SectionViewHolder(view)
    }

    override fun onBindViewHolder(holder: SectionViewHolder, position: Int) {
        val section = sections[position]

        holder.textTitle.text = section.title

        val itemAdapter = ItemListAdapter(section, itemClickListener)
        holder.recyclerView.adapter = itemAdapter
        holder.recyclerView.setHasFixedSize(false)
        holder.recyclerView.clearItemDecorations()
        holder.recyclerView.addItemDecoration(SpaceItemDecoration(
            activity.resources.getDimensionPixelSize(R.dimen.card_horizontal_margin) / 2,
            activity.resources.getDimensionPixelSize(R.dimen.card_vertical_margin) / 2,
            false,
            object : SpaceItemDecoration.RecyclerViewInfo {
                override fun isHeader(position: Int): Boolean {
                    return false
                }

                override val spanCount: Int
                    get() = holder.recyclerView.spanCount

                override val itemCount: Int
                    get() = itemAdapter.itemCount
            }
        ))
        ViewCompat.setNestedScrollingEnabled(holder.recyclerView, false)

        if (section.hasAccessibleItems()) {
            contentAvailable(section, holder)
            itemAdapter.updateItems(section.accessibleItems)
        } else {
            contentLocked(section, holder)
        }
    }

    private fun contentAvailable(section: Section, holder: SectionViewHolder) {
        holder.progressBar.visibility = View.GONE
        holder.viewModuleNotification.visibility = View.GONE
        holder.viewHeader.setBackgroundColor(
            ContextCompat.getColor(
                activity,
                R.color.section_header_bg
            )
        )
        holder.textTitle.setTextColor(ContextCompat.getColor(activity, R.color.section_header_text))

        val outValue = TypedValue()
        activity.theme.resolveAttribute(android.R.attr.selectableItemBackground, outValue, true)
        holder.layout.foreground = ContextCompat.getDrawable(activity, outValue.resourceId)
        holder.layout.setOnClickListener { sectionClickListener.onSectionClicked(section.id) }

        if (section.hasDownloadableContent()) {
            holder.viewDownloadButton.visibility = View.VISIBLE
            holder.viewDownloadButton.setOnClickListener {
                sectionClickListener.onSectionDownloadClicked(
                    section.id
                )
            }
        } else {
            holder.viewDownloadButton.visibility = View.GONE
        }
    }

    private fun contentLocked(section: Section, holder: SectionViewHolder) {
        holder.progressBar.visibility = View.GONE
        holder.viewModuleNotification.visibility = View.VISIBLE
        holder.viewHeader.setBackgroundColor(
            ContextCompat.getColor(
                activity,
                R.color.section_header_locked_bg
            )
        )
        holder.textTitle.setTextColor(
            ContextCompat.getColor(
                activity,
                R.color.section_header_locked_text
            )
        )

        holder.layout.isClickable = false
        holder.layout.foreground = null
        holder.viewDownloadButton.visibility = View.GONE

        if (section.startDate != null && section.startDate.isFuture) {
            val dateOut = if (activity.is7inchTablet) {
                DateFormat.getDateTimeInstance(
                    DateFormat.LONG,
                    DateFormat.SHORT,
                    Locale.getDefault()
                )
            } else {
                DateFormat.getDateTimeInstance(
                    DateFormat.MEDIUM,
                    DateFormat.SHORT,
                    Locale.getDefault()
                )
            }

            holder.textModuleNotification.text = String.format(
                activity.getString(R.string.available_at),
                dateOut.format(section.startDate!!)
            )
            holder.textModuleNotification.setDate(section.startDate!!)
        } else {
            holder.textModuleNotification.text =
                activity.getString(R.string.module_notification_no_content)
        }
    }

    interface OnSectionClickListener {

        fun onSectionClicked(sectionId: String)

        fun onSectionDownloadClicked(sectionId: String)
    }

    class SectionViewHolder(view: View) : RecyclerView.ViewHolder(view) {

        @BindView(R.id.container)
        lateinit var layout: FrameLayout

        @BindView(R.id.textTitle)
        lateinit var textTitle: TextView

        @BindView(R.id.recyclerView)
        lateinit var recyclerView: AutofitRecyclerView

        @BindView(R.id.containerProgress)
        lateinit var progressBar: ProgressBar

        @BindView(R.id.header)
        lateinit var viewHeader: View

        @BindView(R.id.moduleNotificationContainer)
        lateinit var viewModuleNotification: View

        @BindView(R.id.moduleNotificationLabel)
        lateinit var textModuleNotification: DateTextView

        @BindView(R.id.downloadBtn)
        lateinit var viewDownloadButton: View

        init {
            ButterKnife.bind(this, view)
        }
    }
}
